bundles/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/SerialPIMHandler.java

```diff
@@ -0,0 +1,179 @@
+        if (futSerialPortInit != null) {
+            futSerialPortInit.cancel(true);
+            this.futSerialPortInit = null;
+        }
+        final SerialIoThread receiveThread = this.receiveThread;
+        if (receiveThread != null) {
+            receiveThread.terminate();
+            try {
+                receiveThread.join(1000);
+            } catch (final InterruptedException e) {
+                // ignore
+            }
+            this.receiveThread = null;
+        }
+        logger.debug("Stopped UPB serial handler");
+        super.dispose();
+    }
+
+    private void openSerialPort(final String portId) {
+        try {
+            final SerialPort serialPort = tryOpenSerialPort(portId);
+            if (serialPort == null) {
+                futSerialPortInit = scheduler.schedule(() -> openSerialPort(portId), SERIAL_PORT_OPEN_RETRY_DELAY_MS,
+                        TimeUnit.MILLISECONDS);
+                return;
+            }
+            logger.debug("Starting receive thread");
+            final SerialIoThread receiveThread = new SerialIoThread(serialPort, this);
+            this.receiveThread = receiveThread;
+            // Once the receiver starts, it may set the PIM status to ONLINE
+            // so we must ensure all initialization is finished at that point.
+            receiveThread.start();
+            updateStatus(ThingStatus.ONLINE);
+        } catch (final Exception e) {
+            logger.warn("failed to open serial port", e);
+        }
+    }
+
+    private @Nullable SerialPort tryOpenSerialPort(final String portId) {
+        logger.debug("opening serial port {}", portId);
+        final SerialPortIdentifier portIdentifier = serialPortManager.getIdentifier(portId);
+        if (portIdentifier == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    Constants.OFFLINE_SERIAL_EXISTS);
+            return null;
+        }
+
+        final SerialPort serialPort;
+        try {
+            serialPort = portIdentifier.open("org.openhab.binding.upb", 1000);
+        } catch (final PortInUseException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    Constants.OFFLINE_SERIAL_INUSE);
+            return null;
+        }
+        try {
+            serialPort.setSerialPortParams(BAUD_RATE, SerialPort.DATABITS_8, SerialPort.STOPBITS_1,
+                    SerialPort.PARITY_NONE);
+            serialPort.setFlowControlMode(SerialPort.FLOWCONTROL_NONE);
+            try {
+                serialPort.enableReceiveThreshold(1);
+                serialPort.enableReceiveTimeout(SERIAL_RECEIVE_TIMEOUT);
+            } catch (final UnsupportedCommOperationException e) {
+                // ignore - not supported for RFC2217 ports
+            }
+        } catch (final UnsupportedCommOperationException e) {
+            logger.warn("cannot open serial port", e);
```

Comment:
```suggestion
            logger.debug("cannot open serial port", e);
```
Message already set in status, which is also logged.